### PR TITLE
Supports attention_mask with 2D、3D and 4D Tensors

### PR DESCRIPTION
**Pull request by yingyibiao**
### PR types

Function optimization
### PR changes

Models
### Description

Extend attention mask (extend to 4d).

### CODE REVIEW

File: paddlenlp/transformers/albert/modeling.py [hunk 1]
```diff
@@ -1030,10 +1030,8 @@ def forward(
         if token_type_ids is None:
```
Comment (guoshengCS) [inline paddlenlp/transformers/albert/modeling.py:1061]: `attention_mask is None`是否有统一的语义呢，看ALBERT这里是表示不使用mask，BERT、BART、blenderbot里是对input_ids中等于pad_token_id的进行mask

Comment (yingyibiao) [reply paddlenlp/transformers/albert/modeling.py:1061]: 1. 我们目前 `attention_mask is None` 的语义存在两种情况（ ALBERT 是对齐 HF 的，BERT 是没有对齐 HF 的），当前的 PR 没有对这个语义进行统一。
2. 如果我们统一对齐到 HF ，我们需要同步更新 tokenizer 的返回字段（目前 tokenizer 的返回字段还没有完全和 HF 一致）。

File: paddlenlp/transformers/model_utils.py [hunk 2]
```diff
@@ -155,6 +155,35 @@ def get_input_embeddings(self):
...
+        Returns:
+            Tensor: The extended attention mask, with a the same dtype as `attention_mask.dtype`.
+        """
+
+        # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
+        # ourselves in which case we just need to make it broadcastable to all heads.
+        if attention_mask.ndim == 3:
+            attention_mask = attention_mask.unsqueeze(axis=1)
+        # Provided a padding mask of dimensions [batch_size, seq_length]
+        # Todo yingyibiao
```
Comment (guoshengCS) [inline paddlenlp/transformers/model_utils.py]: 注意 TODO 的格式  https://zh-google-styleguide.readthedocs.io/en/latest/google-python-styleguide/python_style_rules/#todo

File: paddlenlp/transformers/model_utils.py [hunk 3]
```diff
@@ -155,6 +155,35 @@ def get_input_embeddings(self):
...
+        # - if the model is a decoder, apply a causal mask in addition to the padding mask
+        # - if the model is an encoder, make the mask broadcastable to
+        # [batch_size, num_heads, seq_length, seq_length]
+        elif attention_mask.ndim == 2:
+            attention_mask = attention_mask.unsqueeze(axis=[1, 2])
+
+        extended_attention_mask = attention_mask.astype(
+            paddle.get_default_dtype())
```
Comment (guoshengCS) [inline paddlenlp/transformers/model_utils.py]: 这里建议和HF一致，使用model中权重的参数，`paddle.get_default_dtype()`不见得和model中权重的dtype是一致的，比如在float32初始化model后调用`model.to('float16')`这样

File: paddlenlp/transformers/bart/modeling.py [hunk 4]
```diff
@@ -211,11 +211,8 @@ def forward(self, input_ids=None, attention_mask=None, **kwargs):
                 input_ids == self.pad_token_id,
                 dtype=paddle.get_default_dtype()).unsqueeze([1, 2]) * -1e4
         # For 2D attention_mask from tokenizer
-        elif attention_mask.ndim == 2:
-            attention_mask = paddle.unsqueeze(
-                attention_mask, axis=[1, 2]).astype(paddle.get_default_dtype())
-            attention_mask = (1.0 - attention_mask) * -1e4
-        attention_mask.stop_gradient = True
+        else:
+            attention_mask = self.get_extended_attention_mask(attention_mask)
```
Comment (guoshengCS) [inline paddlenlp/transformers/bart/modeling.py]: 也再确认下当前这样的话attention_mask是否还是stop_gradient的吧

File: paddlenlp/transformers/bart/modeling.py [hunk 5]
```diff
@@ -306,6 +303,8 @@ def forward(self,
         hidden_states = self.decoder_layernorm_embedding(hidden_states)
         decoder_input = self.decoder_dropout(hidden_states)
 
+        if memory_mask is not None:
```
Comment (guoshengCS) [inline paddlenlp/transformers/bart/modeling.py:304]: 上面`decoder_attention_mask`的处理能否搞成跟HF一致的呢，生成的casual mask 加上传入的 decoder_attention_mask

File: paddlenlp/transformers/bart/modeling.py [hunk 6]
```diff
@@ -507,28 +510,23 @@ def forward(self,
...
-                                      encoder_output, attention_mask, cache)
 
+        if attention_mask is not None:
+            memory_mask = attention_mask
+        elif input_ids is not None:
+            memory_mask = (input_ids != self.pad_token_id)
+        else:
+            memory_mask = None
+
+        decoder_output = self.decoder(decoder_input_ids, decoder_attention_mask,
```
Comment (guoshengCS) [inline paddlenlp/transformers/bart/modeling.py:526]: 这里整体的表现也请 @gongel  确认下，包括训练和预测的，看着我们当前预测时 decoder_attention_mask 处理和HF也不太一样，我们当前`prepare_inputs_for_generation`里还会处理decoder_attention_mask，HF的`prepare_inputs_for_generation`没有处理decoder_attention_mask

File: paddlenlp/transformers/blenderbot/modeling.py [hunk 7]
```diff
@@ -720,13 +723,12 @@ class :class:`paddle.nn.TransformerDecoder` for more information regarding ``cac
...
-                input_ids == self.pad_token_id,
-                dtype=paddle.get_default_dtype()).unsqueeze([1, 2]) * -1e4
-            memory_mask.stop_gradient = True
-        else:
+        if attention_mask is not None:
             memory_mask = attention_mask
+        elif input_ids is not None:
+            memory_mask = (input_ids != self.pad_token_id)
+        else:
+            memory_mask = None
```
Comment (guoshengCS) [inline paddlenlp/transformers/blenderbot/modeling.py]: 这里mask看着是和BART一样的处理，在HF中也是和BART一样的处理。但是我们的BART和Blender实现中 prepare_inputs_for_generation 不一致，BART的prepare_inputs_for_generation中会多一个对decoder_attention_mask的处理，Blender 没有，这个是否会有问题

Comment (guoshengCS) [reply paddlenlp/transformers/blenderbot/modeling.py]: 另外Blender是没有example吗

File: paddlenlp/transformers/ctrl/modeling.py [hunk 8]
```diff
@@ -490,24 +490,12 @@ def forward(self,
...
-            # masked positions, this operation will create a tensor which is 0.0 for
-            # positions we want to attend and -10000.0 for masked positions.
-            # Since we are adding it to the raw scores before the softmax, this is
-            # effectively the same as removing these entirely.
-            attention_mask = attention_mask.astype(
-                dtype=paddle.get_default_dtype())  # fp16 compatibility
-            attention_mask = (1.0 - attention_mask) * -10000.0
+        if attention_mask is None:
+            attention_mask = paddle.cast(
+                input_ids == self.pad_token_id,
```
Comment (guoshengCS) [inline paddlenlp/transformers/ctrl/modeling.py]: 这里 `self.pad_token_id`应该获取不到。看了下这个有demo，如果有example的，一定验证下example的正确性。

File: paddlenlp/transformers/distilbert/modeling.py [hunk 9]
```diff
@@ -257,12 +257,13 @@ def forward(self, input_ids, attention_mask=None):
...
         '''
-
         if attention_mask is None:
-            attention_mask = paddle.unsqueeze(
-                (input_ids == self.pad_token_id
-                 ).astype(self.encoder.layers[0].norm1.weight.dtype) * -1e4,
-                axis=[1, 2])
+            attention_mask = paddle.cast(
+                input_ids == self.pad_token_id,
+                dtype=paddle.get_default_dtype()).unsqueeze([1, 2]) * -1e4
```
Comment (guoshengCS) [inline paddlenlp/transformers/distilbert/modeling.py]: `paddle.get_default_dtype` 不见得是和 weight 的dtype一致的，同样建议类似HF使用weight的dtype，最好整体刷一下

File: paddlenlp/transformers/ernie/modeling.py [hunk 10]
```diff
@@ -644,14 +644,10 @@ def forward(self,
         if attention_mask is None:
             attention_mask = paddle.unsqueeze(
                 (input_ids == self.pad_token_id
-                 ).astype(self.pooler.dense.weight.dtype) * -1e4,
+                 ).astype(paddle.get_default_dtype()) * -1e4,
```
Comment (guoshengCS) [inline paddlenlp/transformers/ernie/modeling.py]: `paddle.get_default_dtype` 不见得是和 weight 的dtype一致的，HF中mask的astype也是使用的weight的dtype

File: paddlenlp/transformers/layoutlmv2/modeling.py [hunk 11]
```diff
@@ -816,11 +817,11 @@ def forward(self,
...
 
         if attention_mask is None:
-            attention_mask = paddle.ones(input_shape)
-
-        visual_attention_mask = paddle.ones(visual_shape)
-
-        attention_mask = attention_mask.astype(visual_attention_mask.dtype)
+            attention_mask = paddle.cast(
+                input_ids != self.pad_token_id,
+                dtype=paddle.get_default_dtype())
```
Comment (guoshengCS) [inline paddlenlp/transformers/layoutlmv2/modeling.py]: 这里看着特殊只支持2D的是吗，也注意文档

File: paddlenlp/transformers/mbart/modeling.py [hunk 12]
```diff
@@ -582,27 +577,23 @@ def forward(self,
...
+
+        if attention_mask is not None:
+            memory_mask = attention_mask
+        elif input_ids is not None:
+            memory_mask = (input_ids != self.pad_token_id)
+        else:
+            memory_mask = None
         decoder_output = self.decoder(decoder_input_ids, decoder_attention_mask,
-                                      encoder_output, attention_mask, cache)
+                                      encoder_output, memory_mask, cache)
```
Comment (guoshengCS) [inline paddlenlp/transformers/mbart/modeling.py]: 同BART

File: paddlenlp/transformers/mobilebert/modeling.py [hunk 13]
```diff
@@ -1161,15 +1161,16 @@ def forward(
                 "You have to specify either input_ids or inputs_embeds")
 
         if attention_mask is None:
-            attention_mask = paddle.ones(input_shape, dtype=input_ids.dtype)
+            extended_attention_mask = paddle.cast(
+                input_ids == self.pad_token_id,
+                dtype=paddle.get_default_dtype()).unsqueeze([1, 2]) * -1e4
```
Comment (guoshengCS) [inline paddlenlp/transformers/mobilebert/modeling.py]: 这里看着 `self.pad_token_id` 也获取不到

File: paddlenlp/transformers/t5/modeling.py [hunk 14]
```diff
@@ -922,8 +922,16 @@ def forward(self,
                 self.is_decoder
             ), f"`use_cache` can only be set to `True` if {self} is used as a decoder"
 
+        # We can provide a self-attention mask of dimensions [batch_size, from_seq_length, to_seq_length]
+        # ourselves in which case we just need to make it broadcastable to all heads.
         if attention_mask is None:
-            attention_mask = paddle.ones(shape=[batch_size, mask_seq_length])
+            extended_attention_mask = paddle.cast(
+                input_ids == self.pad_token_id,
+                dtype=paddle.get_default_dtype()).unsqueeze([1, 2]) * -1e4
```
Comment (guoshengCS) [inline paddlenlp/transformers/t5/modeling.py]: 这里`self.pad_token_id`获取不到，这个有demo，也注意测试

Comment (guoshengCS) [reply paddlenlp/transformers/t5/modeling.py]: 另外这里的处理会稍微复杂些，如果表示的是encoder可以这样，如果表示的是decoder的话默认的不能是这样，需要是三角的，HF中decoder默认也会是三角的

File: paddlenlp/transformers/bart/modeling.py (hunk 6 above)
Comment (gongel) [reply paddlenlp/transformers/bart/modeling.py:526]: @yingyibiao 这里的`memory_mask `可以挪到`if encoder_output is None:`前面去生成吗？这样可以减少一次`attention_mask`的生成，MBart同理

File: paddlenlp/transformers/bart/modeling.py (hunk 4 above)
Comment (yingyibiao) [reply paddlenlp/transformers/bart/modeling.py]: 已再次确认，stop_gradient=True

File: paddlenlp/transformers/albert/modeling.py (hunk 1 above)
Comment (yingyibiao) [reply paddlenlp/transformers/albert/modeling.py:1061]: ALBERT 已经和 BERT 统一

File: paddlenlp/transformers/mobilebert/modeling.py (hunk 13 above)
Comment (yingyibiao) [reply paddlenlp/transformers/mobilebert/modeling.py]: 已修改，设置 self.pad_token_id = pad_token_id

File: paddlenlp/transformers/ernie/modeling.py (hunk 10 above)
Comment (yingyibiao) [reply paddlenlp/transformers/ernie/modeling.py]: 在 PretrainedModel 类中添加了 dtype 方法，默认获取 weight 的 dtype

File: paddlenlp/transformers/layoutlmv2/modeling.py (hunk 11 above)
Comment (yingyibiao) [reply paddlenlp/transformers/layoutlmv2/modeling.py]: 这里因为需要同时构造 attention_mask 和 visual_attention_mask，两者拼接后进行后处理，所以只支持 2D 输入。

File: paddlenlp/transformers/distilbert/modeling.py (hunk 9 above)
Comment (yingyibiao) [reply paddlenlp/transformers/distilbert/modeling.py]: Done

File: paddlenlp/transformers/model_utils.py (hunk 3 above)
Comment (yingyibiao) [reply paddlenlp/transformers/model_utils.py]: Done

File: paddlenlp/transformers/model_utils.py (hunk 2 above)
Comment (yingyibiao) [reply paddlenlp/transformers/model_utils.py]: Todo 已删除

File: paddlenlp/transformers/bart/modeling.py (hunk 5 above)
Comment (yingyibiao) [reply paddlenlp/transformers/bart/modeling.py:304]: Done
分别采用 get_extended_attention_mask 和 get_extended_attention_mask_for_decoder 来处理 encoder 和 decoder 的 attention。

File: paddlenlp/transformers/bart/modeling.py (hunk 6 above)
Comment (yingyibiao) [reply paddlenlp/transformers/bart/modeling.py:526]: > @yingyibiao 这里的`memory_mask `可以挪到`if encoder_output is None:`前面去生成吗？这样可以减少一次`attention_mask`的生成，MBart同理

Done

File: paddlenlp/transformers/blenderbot/modeling.py (hunk 7 above)
Comment (yingyibiao) [reply paddlenlp/transformers/blenderbot/modeling.py]: 已经去除 BART prepare_inputs_for_generation 对 decoder_attention_mask 的处理

Comment (yingyibiao) [reply paddlenlp/transformers/blenderbot/modeling.py]: > 另外Blender是没有example吗

没有 example

File: paddlenlp/transformers/ctrl/modeling.py (hunk 8 above)
Comment (yingyibiao) [reply paddlenlp/transformers/ctrl/modeling.py]: 已设置 self.pad_token_id

File: paddlenlp/transformers/t5/modeling.py (hunk 14 above)
Comment (yingyibiao) [reply paddlenlp/transformers/t5/modeling.py]: > 这里`self.pad_token_id`获取不到，这个有demo，也注意测试

Done

File: paddlenlp/transformers/mbart/modeling.py (hunk 12 above)
Comment (yingyibiao) [reply paddlenlp/transformers/mbart/modeling.py]: Done

**Comment by guoshengCS** [comment]
这个也再解决下冲突更新下吧 @yingyibiao

**Comment by guoshengCS** [comment]
另外当前有测试过哪些呢

File: paddlenlp/transformers/bart/modeling.py [hunk 15]
```diff
@@ -289,21 +282,28 @@ def forward(self,
...
+        batch_size, seq_length = input_shape
+
         past_key_values_length = paddle.shape(
             cache[0][0].k)[2] if cache is not None else 0
+        mask_seq_length = past_key_values_length + seq_length if past_key_values_length != 0 else seq_length
+
+        if decoder_attention_mask is None:
+            decoder_attention_mask = paddle.ones([batch_size, mask_seq_length])
+        decoder_attention_mask = self.get_extended_attention_mask_for_decoder(
+            decoder_attention_mask, input_shape)
```
Comment (guoshengCS) [inline paddlenlp/transformers/bart/modeling.py:295]: 就BART看着decoder这里的mask和HF的是不是还是会不一样，比如HF Bart只会在 input_shape[-1] > 1 的时候才使用causal_mask ，而且若 decoder_attention_mask 是None的话HF Bart中就不加这个了

**Comment by sijunhe** [comment]
Closing at the request of @guoshengCS
